hazelcast/src/main/java/com/hazelcast/jet/impl/AbstractJobProxy.java

```diff
@@ -294,6 +296,33 @@ private void terminate(TerminationMode mode) {
         }
     }
 
+    /**
+     * @implNote Listeners added to a job after it completes will not be removed automatically
+     * since the job has already produced a terminal event. In order to make auto-deregistration
+     * race-free, it is not allowed to add listeners to completed jobs. Checking whether the job
+     * status is terminal before adding the listener will not work since they are not atomic.
+     * Delegating the listener registration to the job coordinator requires locking the master
+     * context, converting the registration operation to an event callback with the same
+     * orderKey as job status events, and restoring Registration#listener during deserialization.
+     * The fastest and simplest solution is to assume optimistically that the job is still
+     * running and remove the listener if it turns out be completed.
+     */
+    @Override
+    public UUID addStatusListener(@Nonnull JobStatusListener listener) {
+        UUID registrationId = doAddStatusListener(listener);
+        String error;
+        try {
+            JobStatus status = getStatus();
```

Comment:
a race is still possible if the job terminates after `doAddStatusListener()` and before  `getStatus()`. Listener will be notified about the event (they may be not only terminal events) even though `addStatusListener` will throw `IllegalStateException`. Or do I miss something and `removeStatusListener` will be executed on the same thread where events for this job are processed?

anyway, this race does not seem so important